plugins/obs-outputs/rtmp-stream.c

```diff
@@ -1065,9 +1119,166 @@ static void check_to_drop_frames(struct rtmp_stream *stream, bool pframes)
+	int i_nal_hrd = strcmp(encoder_id, "obs_x264") == 0 ?
+			obs_data_get_int(params, "i_nal_hrd"): 0;
+
+	uint64_t cur_time_ms = os_gettime_ns() / 1000000;
+	uint64_t last_adjustment_time = stream->last_adjustment_time;
+
+	size_t counter = stream->congestion_counter;
+	stream->mean_congestion += (stream->congestion
+			- stream->congestion_array[counter]) / CONGESTION_ARRAY_SIZE;
+	stream->congestion_array[counter] = stream->congestion;
+	stream->congestion_counter = (stream->congestion_counter + 1) %
+			CONGESTION_ARRAY_SIZE;
+	float congestion = stream->congestion;
+	float last_congestion = stream->last_congestion;
+	float mean_congestion = stream->mean_congestion;
+	float max_congestion = find_maximum(stream->congestion_array,
+			CONGESTION_ARRAY_SIZE);
+
+	int current_bitrate = stream->dynamic_bitrate;
+	int initial_bitrate = stream->initial_bitrate;
+	int previous_bitrate;
+
+	float decrease_rate = (float)stream->bitrate_decrease_rate;
+	float increase_rate = (float)stream->bitrate_increase_rate;
+	float dynamic_threshold = 0.01 * (float)stream->dynamic_threshold;
+	uint64_t polling_time = 1000 * stream->recovery_polling_time; // in millisecs
+	uint64_t decrease_polling_time = stream->decrease_polling_time; // set in millisec in UI
+
+	/* X264_NAL_HRD_CBR=2 incompatible with dynamic variable bitrate.
+	 * On default settings, bitrate is adjusted downwards every second by 
+	 * about 15%.
+	 * Detection threshold of congestion is 0.15 on default but could be larger;
+	 * this would mean less reactivity but would allow micro-congestions
+	 * to heal themselves without changing too soon the bitrate.
+	 * We define a minimal bitrate at half the initial one.
+	 * The polling time tells how many milliseconds one waits before trying
+	 * to increase the bitrate.
+	 * The larger the polling time, the lower you want the threshold to be
+	 * in order to counter-balance the potential of many drops.
+	 * You also want the increase rate to be set at larger values.
+	 * Ex: if polling_time : 5 sec -> 30 sec
+	 * increase_rate : 10 % -> 30 - 40 %
+	 * threshold: 15 % congestion -> 5 %
+	 * decrease_polling_time: 1000 millisec -> 300 millisec
+	 */
+	bool decrease_br = congestion > dynamic_threshold &&
+			current_bitrate > (int)(initial_bitrate / 2) &&
+			(cur_time_ms - last_adjustment_time) > decrease_polling_time;
+
+	bool increase_br = mean_congestion < 0.05 && max_congestion < 0.15 &&
+			congestion < 0.05 && last_congestion < 0.05 &&
+			current_bitrate < initial_bitrate &&
+			(cur_time_ms - last_adjustment_time) > polling_time;
+
+	bool revert_increase_br = congestion > 0.01 && stream->increase_just_attempted;
+	if ((cur_time_ms - last_adjustment_time) > decrease_polling_time)
+		stream->increase_just_attempted = false;
+
+	if (i_nal_hrd != 2 && stream->switch_variable_bitrate) {
+		/* Bitrate is adjusted downwards by 15% every second (on default
+		 * settings). The bitrate floor is arbitrarily set at 50 % of initial
```

Comment:
I have build and tested this under my own conditions and have concluded that the bitrate floor should be user configurable. (It's the only problem I had with this) In some conditions for me the bitrate can vary more than 50%. Personally I would like to set my bitrate at 5000kb/s However, under heavy congestion I may only be able to achieve 1000kb/s even only 250kb/s at the worst times. I know the quality will be bad however I still prefer that than dropping frames. Other reasons is low end streams with an initial bitrate of 1000kb/s will only drop to 500kb/s under some conditions available upload may only be 200kb/s

I suggest making this user configurable as it can be very important in some cases.

Reply:
I have no objection to letting the user set the floor bitrate.
50 % should be enough in most cases though because the resolution is not changed so lowering the bitrate even more is kind of violent. But I admit that some prefer to lower quality all the way rather than dropping frames.
Did you test several encoders? Qsv behaves better than the others.

Reply:
Yeah I get you with not being able to change resolution. Just the odd times my upload does fall under 50% the bitrate floor it would be nice just to show something that looks a little bad as it's better than nothing by a lot. When I'm streaming to twitch at 6000kb/s bitrate the video can still look good at 1500kb/s for some less complex parts as that's usually my go to bit-rate when my upload is bad at rare points.

The main point I see from dynamic bitrate is just the goal of not dropping frames but still pushing the best quality the upload can handle. The bitrate floor can sometimes be a limiting factor for this and even if it falls below 50% I'm sure people would prefer the lower quality than dropping frames (Is there any benefit to dropping frames?) Also I couldn't get to test other encoders as I don't have QSV.

Reply:
I'll add the floor setting next week along with other things

Reply:
I'll add the floor setting next week along with other things

Reply:
Thanks :) This PR was the only reason I went to find out how to build OBS since for me and others it can be such an important feature.

Reply:
bitrate floor setting added.